### PR TITLE
karmada operator: add deinit workflow

### PR DESCRIPTION
**Pull request by calvin0327**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
add karmada operator deInit workflow to remove a karmada instance. the workflow only have three task: `RemoveComponent` => `CleanupCert` => `CleanupKubeconfig`.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
Test result:

image

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by calvin0327** [comment]
@lonelyCZ @Poor12 PTAL

File: operator/pkg/controller/karmada/controller.go [hunk 1]
```diff
@@ -58,41 +58,31 @@ func (ctrl *Controller) Reconcile(ctx context.Context, req controllerruntime.Req
 		if !controllerutil.ContainsFinalizer(karmada, ControllerFinalizerName) {
 			controllerutil.AddFinalizer(karmada, ControllerFinalizerName)
 			if err := ctrl.Update(ctx, karmada); err != nil {
-				return controllerruntime.Result{}, err
+				return controllerruntime.Result{Requeue: true}, err
```
Comment (Poor12) [inline operator/pkg/controller/karmada/controller.go:61]: I remember the controller will requeue the request to be processed if err is not nil https://github.com/karmada-io/karmada/pull/2973#discussion_r1052886320. I feel like add or not-add is both ok. The key is that we might need to unify the code style in one module which may not confuse others.

File: operator/pkg/tasks/deint/component.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+)
+
+// NewRemoveComponentTask init a remove karmada components task
+func NewRemoveComponentTask() workflow.Task {
+	return workflow.Task{
+		Name:        "remove-component",
+		Run:         runRemoveComponent,
+		RunSubTasks: true,
+		Tasks: []workflow.Task{
+			newRemoveComponentSubTask(constants.KarmadaWebhookComponent, util.KarmadaWebhookName),
```
Comment (Poor12) [inline operator/pkg/tasks/deint/component.go:22]: I remember there is some svc installed. Do we need to uninstall them when deinit?

File: operator/pkg/controller/karmada/controller.go (hunk 1 above)
Comment (calvin0327) [reply operator/pkg/controller/karmada/controller.go:61]: I agree to that

File: operator/pkg/tasks/deint/component.go (hunk 2 above)
Comment (calvin0327) [reply operator/pkg/tasks/deint/component.go:22]: I want to set the service OwnReference to their workflow. so, we only delete the worklfow.
@Poor12  How do you think?

Comment (Poor12) [reply operator/pkg/tasks/deint/component.go:22]: For now I think it's okay. I wonder if there is a scenario where we need to consider deployment and svc separately.

Comment (calvin0327) [reply operator/pkg/tasks/deint/component.go:22]: Ok, If the service resource exists alone, it doesn't to make sense. I seem not encounter the scenario.

**Comment by lonelyCZ** [comment]
Thanks @calvin0327 , I will test it in my env ASAP.

/assign

**Comment by Poor12** [comment]
Basicly LGTM
Leave comments for @lonelyCZ .

File: operator/pkg/controller/karmada/controller.go [hunk 3]
```diff
@@ -61,38 +61,28 @@ func (ctrl *Controller) Reconcile(ctx context.Context, req controllerruntime.Req
...
 		return controllerruntime.Result{}, err
 	}
-	return planner.Execute()
-}
+	if err := planner.Execute(); err != nil {
+		return controllerruntime.Result{}, err
+	}
+
+	// The object is being deleted
+	if !karmada.DeletionTimestamp.IsZero() && controllerutil.ContainsFinalizer(karmada, ControllerFinalizerName) {
```
Comment (lonelyCZ) [inline operator/pkg/controller/karmada/controller.go:76]: Suggest to move this forward to keep the controller code style. Firstly check `!karmada.DeletionTimestamp.IsZero()` and then sync cr, which can reduce the number of judgments.

like 

https://github.com/karmada-io/karmada/blob/a4d3eced85ede213c9c6cf6813ea564dde1b6583/pkg/controllers/cluster/cluster_controller.go#L171-L175

File: operator/pkg/controller/karmada/controller.go [hunk 4]
```diff
@@ -61,38 +61,28 @@ func (ctrl *Controller) Reconcile(ctx context.Context, req controllerruntime.Req
...
+		// remove our finalizer from the list and update it.
```
Comment (lonelyCZ) [inline operator/pkg/controller/karmada/controller.go:77]:
```suggestion
		// remove the finalizer from the list and update it.
```

File: operator/pkg/controller/karmada/controller.go [hunk 5]
```diff
@@ -61,38 +61,28 @@ func (ctrl *Controller) Reconcile(ctx context.Context, req controllerruntime.Req
...
+		controllerutil.RemoveFinalizer(karmada, ControllerFinalizerName)
```
Comment (lonelyCZ) [inline operator/pkg/controller/karmada/controller.go:78]: Suggest to encapsulate it as a function.

like

https://github.com/karmada-io/karmada/blob/a4d3eced85ede213c9c6cf6813ea564dde1b6583/pkg/controllers/execution/execution_controller.go#L171-L183

Comment (calvin0327) [reply operator/pkg/controller/karmada/controller.go:85]: ok

File: operator/pkg/controller/karmada/controller.go (hunk 3 above)
Comment (calvin0327) [reply operator/pkg/controller/karmada/controller.go]: @lonelyCZ  Ok, I got it, but in this scenario, must put the check `!cluster.DeletionTimestamp.IsZero()`  behind the process of `reconciliation` due to both Init and Deinit workflow are to excute the following logic:

```golang
klog.V(2).InfoS("Reconciling karmada", "name", req.Name)
	planner, err := NewPlannerFor(karmada, ctrl.Client, ctrl.Config)
	if err != nil {
		return controllerruntime.Result{}, err
	}
	if err := planner.Execute(); err != nil {
		return controllerruntime.Result{}, err
	}
```

**Comment by lonelyCZ** [comment]
Hi @calvin0327 , could you provide `operator/config/deploy/karmada-operator.yaml` again, I want to test it in my env.

**Comment by calvin0327** [comment]
> Hi @calvin0327 , could you provide `operator/config/deploy/karmada-operator.yaml` again, I want to test it in my env.

Of cource, here:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: karmada-operator
  namespace: karmada-system
  labels:
    karmada-app: karmada-operator
spec:
  replicas: 1
  selector:
    matchLabels:
      karmada-app: karmada-operator
  template:
    metadata:
      labels:
        karmada-app: karmada-operator
    spec:
      containers:
      - name: karmada-operator
        image: release.daocloud.io/karmada/karmada-operator:v1.1.0-1965-g20c7316c
        imagePullPolicy: IfNotPresent
        command:
        - /bin/karmada-operator
        - --kubeconfig=/etc/config
        - --v=4
        volumeMounts:
        - name: kubeconfig
          mountPath: /etc/config
          subPath: config
      volumes:
      - name: kubeconfig
        secret:
          secretName: my-kubeconfig
```

**Comment by lonelyCZ** [comment]
Hi @calvin0327 , Is `release.daocloud.io/karmada/karmada-operator:v1.1.0-1965-g10323eac` a newest image?

**Comment by calvin0327** [comment]
> 

The image release.daocloud.io/karmada/karmada-operator:v1.1.0-1965-g10323eac is latest image.

File: operator/pkg/tasks/deint/cert.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
+package tasks
```
Comment (lonelyCZ) [inline operator/pkg/tasks/deint/cert.go:1]: Please update `deint/cert.go` to `deinit/cert.go`

**Review by lonelyCZ — COMMENTED** [review]
I just tested it in my env that worked fine! 

There is only a spell-error here.

**Comment by Poor12** [comment]
> > 
> 
> @lonelyCZ The image release.daocloud.io/karmada/karmada-operator:v1.1.0-1965-g10323eac is latest image.

BTW, we can make image locally. I guess it's the same.

**Comment by lonelyCZ** [comment]
> BTW, we can make image locally. I guess it's the same.

Yes, we can add it to Makefile in next PR.

**Comment by Poor12** [comment]
> > BTW, we can make image locally. I guess it's the same.
> 
> Yes, we can add it to Makefile in next PR.

It's alreay done. You can use `make image-karmada-operator` to make the image.

**Comment by RainbowMango** [comment]
Seems we don't have the image registry on the Docker Hub yet.  
Can we create the registry now?

**Comment by RainbowMango** [comment]
Oh, i see #3270.
I'll get back to #3270 once I created the repo.

File: operator/pkg/tasks/deint/cert.go (hunk 6 above)
Comment (calvin0327) [reply operator/pkg/tasks/deint/cert.go]: updated. @lonelyCZ

**Comment by calvin0327** [comment]
@Poor12 @lonelyCZ @RainbowMango any others coments?

File: operator/pkg/constants/constants.go [hunk 7]
```diff
@@ -76,4 +82,14 @@ const (
 	KarmadaControllerManagerComponent = "KarmadaControllerManager"
 	// KarmadaSchedulerComponent defines the name of karmada scheduler component
 	KarmadaSchedulerComponent = "KarmadaScheduler"
+	// KarmadaWebhookComponent defines the name of the karmada-webhook component
+	KarmadaWebhookComponent = "karmadaWebhook"
```
Comment (lonelyCZ) [inline operator/pkg/constants/constants.go:86]: Is it right?
```suggestion
	KarmadaWebhookComponent = "KarmadaWebhook"
```

Comment (calvin0327) [reply operator/pkg/constants/constants.go]: @lonelyCZ Thanks. done

**Review by RainbowMango — APPROVED** [review]
/approve
Thanks.

**Pull request merged**